### PR TITLE
Convert more node accessors to pybind11

### PR DESCRIPTION
**Pull request by gbalke**
Part of #665.

Handles:
```
rclpy_get_node_logger_name
rclpy_get_node_names_and_namespaces
rclpy_get_node_names_and_namespaces_with_enclaves
rclpy_node_get_fully_qualified_name
```

### CODE REVIEW

**Comment by gbalke** [comment]
CI up to `rclpy`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14017/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8839/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11711/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14102/)

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:44]: @gbalke why change the API? I see the previous implementation used to return `None` (which you can do by changing the return value to `py::object` and returning `py::none()` in this case). Same elsewhere.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:130]: @gbalke nit: 
```suggestion
  for (size_t idx = 0; idx < node_names.size; ++idx) {
```

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:144]: @gbalke nit: 
```suggestion
  rcutils_ret_t rcutils_ret = rcutils_string_array_fini(&node_names);
```

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:155]: @gbalke consider putting all these "fini" calls in an `RCPPUTILS_SCOPE_EXIT` block *before* populating the list, in case something throws while doing so.

Comment (hidmic) [reply rclpy/src/rclpy/node.cpp:155]: Along that same line, make sure to finalize everything before throwing.

File: rclpy/src/rclpy/node.cpp
Comment (gbalke) [reply rclpy/src/rclpy/node.cpp:44]: I was doing this given @sloretz commentary https://github.com/ros2/rclpy/pull/719#discussion_r596448711

File: rclpy/src/rclpy/node.cpp
Comment (gbalke) [reply rclpy/src/rclpy/node.cpp:155]: If I put it in before populating the list then won't the data be destroyed? Or am I missunderstanding?

Comment (gbalke) [reply rclpy/src/rclpy/node.cpp]: > Along that same line, make sure to finalize everything before throwing from this block.

Good call.

Comment (gbalke) [reply rclpy/src/rclpy/node.cpp]: @hidmic I have updated to account for these issues.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [reply rclpy/src/rclpy/node.cpp:46]: Oh, I see. Nevermind my comment then.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:131]: @gbalke nit: no need to have these declared outside the `RCPPUTILS_SCOPE_EXIT`.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:147]: @gbalke this block should go before populating the list, around line 128. [`RCPPUTILS_SCOPE_EXIT`](https://github.com/ros2/rcpputils/blob/master/include/rcpputils/scope_exit.hpp) block will be executed upon leaving the scope it was defined in.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp:158]: @gbalke if an error was set by function A, resetting it before calling function B avoids warnings about error overwrite if function B happens to set an error as well. That is, consider moving each of these `if` clauses after the corresponding function call. It's pure printing, so no risk of an exception getting thrown there.

Comment (gbalke) [reply rclpy/src/rclpy/node.cpp:158]: I think I got mixed up because of https://github.com/ros2/rclpy/pull/730#discussion_r599111444. I'll change it back since I'm not throwing now.

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [inline rclpy/src/rclpy/node.cpp]: @gbalke this won't have the intended effect. If it throws, it'll throw from a destructor and C++ will terminate the program :)

If I may, you may want to do something like:

```c++
RCPPUTILS_SCOPE_EXIT({  // setup cleanup block to be run on scope exit
  rcutils_ret_t fini_ret = rcutils_string_array_fini(&node_names);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini node names during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
  fini_ret = rcutils_string_array_fini(&node_namespaces);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini node namespaces during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
  fini_ret = rcutils_string_array_fini(&enclaves);
  if (RCUTILS_RET_OK != fini_ret) {
    RCUTILS_SAFE_FWRITE_TO_STDERR(
      "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
      "failed to fini enclaves string array during error handling: ");
    RCUTILS_SAFE_FWRITE_TO_STDERR(rcl_get_error_string().str);
    RCUTILS_SAFE_FWRITE_TO_STDERR("\n");
    rcl_reset_error();
  }
});

py::list pynode_names_and_namespaces(node_names.size);
for (size_t idx = 0; idx < node_names.size; ++idx) {
  if (get_enclaves) {
    pynode_names_and_namespaces[idx] = py::make_tuple(
      py::str(node_names.data[idx]),
      py::str(node_namespaces.data[idx]),
      py::str(enclaves.data[idx]));
  } else {
    pynode_names_and_namespaces[idx] = py::make_tuple(
      py::str(node_names.data[idx]),
      py::str(node_namespaces.data[idx]));
  }
}

return pynode_names_and_namespaces;  // and then cleanup code is run automatically
```

**Review by hidmic — COMMENTED** [review]
LGTM but for one comment, pending green CI

File: rclpy/src/rclpy/node.cpp
Comment (gbalke) [reply rclpy/src/rclpy/node.cpp]: What I guess i don't understand is that the fini calls will happen before the for loop? Or is that the special magic of the `SCOPE_EXIT` statement? Where that will be called upon completion?

Comment (gbalke) [reply rclpy/src/rclpy/node.cpp]: Read the comments on the blocks and now it makes sense :+1:. Thanks for walking me through it haha.

**Comment by gbalke** [comment]
CI up to `rclpy`
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14047/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8868/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11740/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14138/)

File: rclpy/src/rclpy/node.cpp
Comment (hidmic) [reply rclpy/src/rclpy/node.cpp]: > Or is that the special magic of the SCOPE_EXIT statement?

If you look under the hood, this `RCPPUTILS_SCOPE_EXIT` macro instantiates an object that takes your code, wraps it in a lambda function, and executes it on _destruction_. Using object construction/destruction to manage resource lifecycles is so common in C++ that [it has a name](https://en.cppreference.com/w/cpp/language/raii#:~:text=Resource%20Acquisition%20Is%20Initialization%20or,in%20limited%20supply)%20to%20the). It's the primary idiom when it comes to exception safety.

**Review by hidmic — APPROVED** [review]

**Comment by gbalke** [comment]
@sloretz lmk if you want to review this before I merge.

**Comment by gbalke** [comment]
@ros-pull-request-builder retest this please :heart:

**Comment by gbalke** [comment]
Ci up to `rclpy`
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14060/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8880/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11752/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14151/)

**Comment by gbalke** [comment]
@ros-pull-request-builder retest this please :broken_heart:

**Pull request merged**
